embabel-common-ai/src/main/kotlin/com/embabel/common/ai/converters/streaming/support/StreamingUtils.kt

```diff
@@ -0,0 +1,232 @@
+            val hasStartTag = startTags.any { line.contains(it) }
+            val hasEndTag = endTags.any { line.contains(it) }
+
+            val state = when {
+                hasStartTag && hasEndTag -> ThinkingState.BOTH
+                hasStartTag -> ThinkingState.START
+                hasEndTag -> ThinkingState.END
+                line.startsWith("//THINKING:") -> ThinkingState.BOTH  // Legacy format
+                else -> ThinkingState.CONTINUATION
+            }
+
+            logger.debug(
+                "Complete pattern classification: hasStart={}, hasEnd={}, state={}",
+                hasStartTag,
+                hasEndTag,
+                state
+            )
+            return state
+        }
+
+        // Check for partial tags (opening or closing tags only)
+        when {
+            startTags.any { trimmed == it } -> {
+                logger.debug("Line is standalone start tag")
+                return ThinkingState.START
+            }
+
+            endTags.any { trimmed == it } -> {
+                logger.debug("Line is standalone end tag")
+                return ThinkingState.END
+            }
+
+            startTags.any { trimmed.startsWith(it) && !endTags.any { end -> trimmed.contains(end) } } -> {
+                logger.debug("Line starts with opening tag but no closing tag")
+                return ThinkingState.START
+            }
+
+            endTags.any { trimmed.endsWith(it) && !startTags.any { start -> trimmed.contains(start) } } -> {
+                logger.debug("Line ends with closing tag but no opening tag")
+                return ThinkingState.END
+            }
+        }
+
+        // No thinking markers found - classify based on content type
+        val isJson = isValidJson(line)
+        return if (isJson) {
+            logger.debug("Line classified as JSON content")
+            ThinkingState.NONE  // This is JSON content, not thinking
+        } else {
+            logger.debug("Line classified as thinking continuation")
+            ThinkingState.CONTINUATION  // Raw text, assume thinking continuation
+        }
+    }
+
+    /**
+     * Fast heuristic validation for JSON content to avoid expensive parsing.
+     *
+     * Performance optimization using simple string checks rather than full JSON parsing
+     * to quickly identify potential JSON lines in streaming content.
+     *
+     * Validation logic:
+     * 1. Line must not be empty after trimming
+     * 2. Must start with '{' (JSON object start)
+     * 3. Must end with '}' (JSON object end)
+     * 4. Must contain ':' (JSON objects require key-value pairs)
+     *
+     * Note: Intentionally simple heuristic - actual JSON parsing in the converter
+     * will catch real JSON syntax errors.
+     *
+     * @param line The line to check for JSON format
+     * @return true if line appears to be JSON object, false otherwise
+     */
+    fun isValidJson(line: String): Boolean {
```

Comment:
This method is only used by this utility itself, and probably should be `private`. It's not a big deal since the whole type is internal anyway.
However, by making the method private, it would even make more sense to rename this utility to `ThinkingDetector`.